lib/preload.js

```diff
@@ -1,32 +1,125 @@
+const SPLASH_IMAGE_PATH_IN_DOCKER = '/img/resin-logo.png'
+
 /** @const {String} Default container name */
 preload.CONTAINER_NAME = 'resin-image-preloader'
 
 /**
  * Build the preloader docker image
- * @param {Object} [processOptions] - child process options
- * @returns {ChildProcess}
+ * @returns Promise
  */
-preload.build = function (processOptions) {
-  processOptions = Object.assign({
-    cwd: process.cwd(),
-    env: process.env,
-    stdio: 'pipe',
-    detached: false
-  }, processOptions)
-
-  const dirname = path.join(__dirname, '..')
-  const argv = [ 'build', '-t', 'resin/resin-preload', dirname ]
-
-  return childProcess.spawn('docker', argv, processOptions)
+preload.build = function () {
+  console.log('Building preloader image...')
+  return docker.buildImage(
+    tarfs.pack(path.resolve(__dirname, '..')),
+    {t: DOCKER_IMAGE_TAG}
+  )
+  .then((stream_) => {
+    return new Promise((resolve, reject) => {
+      docker.modem.followProgress(
+        stream_,
+        (err, output) => {
+          if (err) {
+            reject(err)
+            return
+          }
+          resolve()
+        },
+        (event) => {
+          if (event.stream) {
+            process.stdout.write(event.stream)
+          }
+        }
+      )
+    })
+  })
+}
+
+class DevNull extends stream.Writable {
+  _write (chunk, enc, next) {
+    next()
+  }
+}
+
+class BufferBackedWritableStream extends stream.Writable {
+  constructor () {
+    super(arguments)
+    this.chunks = []
+  }
+
+  _write (chunk, enc, next) {
+    this.chunks.push(chunk)
+    next()
+  }
+
+  getData () {
+    return Buffer.concat(this.chunks)
+  }
+}
+
+const bindMount = (source, target) => {
+  return {
+    Source: path.resolve(source),
+    Target: target,
+    Type: 'bind',
+    Consistency: 'delegated'
+  }
+}
+
+const createContainerDisposer = (options) => {
+  const mounts = [bindMount(options.image, DISK_IMAGE_PATH_IN_DOCKER)]
+  if (options.splashImage) {
+    mounts.push(bindMount(options.splashImage, SPLASH_IMAGE_PATH_IN_DOCKER))
+  }
+  return docker.createContainer({
+    Image: DOCKER_IMAGE_TAG,
+    Name: options.containerName || preload.CONTAINER_NAME,
+    NetworkMode: 'host',
+    Privileged: true,
+    AttachStdout: true,
+    AttachStderr: true,
+    Env: [
+      `COMMAND=${options.command}`,
+      `APP_ID=${options.appId}`,
+      `API_TOKEN=${options.apiToken || ''}`,
+      `API_KEY=${options.apiKey || ''}`,
+      `COMMIT=${options.commit || ''}`,
+      `REGISTRY_HOST=${options.registryHost || ''}`,
+      `API_HOST=${options.apiHost || ''}`,
+      `DONT_DETECT_FLASHER_TYPE_IMAGES=${options.dontDetectFlasherTypeImages ? 'TRUE' : 'FALSE'}`
+    ],
+    Mounts: mounts
+  })
+  .disposer((container) => {
+    return container.remove()
+  })
+}
+
+const runContainer = (container) => {
+  return container.start()
+  .then((container) => {
+    return container.attach({stream: true, stdout: true, stderr: true})
+  })
+  .then((stream_) => {
```

Comment:
Why are you using `stream_` for a variable name? It seems like if you're avoiding name clashes, it would be better to use something like `dockerStream` or `buildStream`.